gcc/rust/backend/rust-compile-expr.cc

```diff
@@ -958,12 +958,71 @@ CompileExpr::visit (HIR::LiteralExpr &expr)
     }
 }
 
+bool
+lvalue_p (const_tree ref)
+{
+  const enum tree_code code = TREE_CODE (ref);
+
+  switch (code)
+    {
+    case REALPART_EXPR:
+    case IMAGPART_EXPR:
+    case COMPONENT_REF:
+      return lvalue_p (TREE_OPERAND (ref, 0));
+
+    case C_MAYBE_CONST_EXPR:
+      return lvalue_p (TREE_OPERAND (ref, 1));
+
+    case COMPOUND_LITERAL_EXPR:
+    case STRING_CST:
+    case CONST_DECL:
+    case INTEGER_CST:
+      return true;
+
+    case MEM_REF:
+    case TARGET_MEM_REF:
+      /* MEM_REFs can appear from -fgimple parsing or folding, so allow them
+	 here as well.  */
+    case INDIRECT_REF:
+    case ARRAY_REF:
+    case VAR_DECL:
+    case PARM_DECL:
+    case RESULT_DECL:
+    case ERROR_MARK:
+      return (TREE_CODE (TREE_TYPE (ref)) != FUNCTION_TYPE
+	      && TREE_CODE (TREE_TYPE (ref)) != METHOD_TYPE);
+
+    case BIND_EXPR:
+      return TREE_CODE (TREE_TYPE (ref)) == ARRAY_TYPE;
+    case PLUS_EXPR:
+    case MINUS_EXPR:
+    case MULT_EXPR:
+    case POINTER_PLUS_EXPR:
+    case POINTER_DIFF_EXPR:
+    case MULT_HIGHPART_EXPR:
+    case TRUNC_DIV_EXPR:
+      return false;
+    default:
+      rust_debug ("unknown");
+      return false;
+    }
+}
+
 void
 CompileExpr::visit (HIR::AssignmentExpr &expr)
 {
   auto lvalue = CompileExpr::Compile (expr.get_lhs (), ctx);
   auto rvalue = CompileExpr::Compile (expr.get_rhs (), ctx);
 
+  if (!lvalue_p (lvalue)
+      || expr.get_lhs ().get_expression_type ()
+	   == HIR::Expr::ExprType::Operator)
+    {
+      rust_error_at (expr.get_lhs ().get_locus (),
```

Comment:
you need to add in the rust error code ussing the proper interface

Reply:
Can you give me more clarity

Reply:
you need to pass the rust error code into this function like we do in other places take a look through the other parts of the code to check